library/src/G4LatticeLogical.cc

```diff
@@ -690,6 +691,148 @@ void G4LatticeLogical::AddValley(G4double phi, G4double theta, G4double psi) {
   fValleyAxis.push_back(fValleyInv.back()*G4ThreeVector(1.,0.,0.));
 }
 
+// Store drifting-electron valley using valley's direction
+
+void G4LatticeLogical::AddValley(const G4ThreeVector& valleyDirVec) {
+      
+  // Find the rotation matrix elements. 
+  // The rows correspond to the rotated axis direction (first row is vx,vy,vz ; second row is a,b,c and third row is d,e,f)
+    
+  // The convention used does not affect HV-transformation. For simplicity, we choose the following conditions :
+  // The rotated y axis must stay in the X-Y plane (c=0)
+  // The rotate Z axis must have positive z component (f>0) 
+  // The three rotated axis should stay normalized : vx^2 + vy^2 + vz^2 = 1 ; a^2 + b^2 = 1 ; d^2 + e^2 + f^2 = 1
+  // And be orthogonal : ad + be = 0 ; avx + bvy = 0 ; dvx + evy + fvz = 0
+  // It must be a right hand coordinate system : (vx,vy,vz) X (a,b,c) = (d,e,f) : d=-vzb ; e=vza ; f=vxb - vya
+    
+  // We first compute f since it needs to be positive and will therefore clear out signs ambiguity 
+  // d^2 + e^2 + f^2 = 1
+  // f^2 = 1 - d^2 - e^2
+  // f^2 = 1 - vz^2 b^2 - vz^2 a^2
+  // f^2 = 1 - vz^2 (b^2 + a^2)
+  // f^2 = 1 - vz^2
+  // f = + sqrt(1 - vz^2)
+    
+    
+  // Then we compute special cases first (vx and vy=0, vx=0 but not vy and vy=0 but not vx).
+    
+  // We start with vx= and vy=0
+  // e=0 and there is only two possible rotation matrix since vz=1 or -1. We manually set the elements to :
+  // b=1 
+  // a=0
+    
+  // If vx=0 but not vy 
+  // a = -e/vy = -sqrt(1-vz^2)/vy = -sqrt(vy^2)/vy = -1 for valleys and 1 for anti-valleys
+  // b=0
+
+  // If vy=0 but not vx 
+  // b = e/vx = sqrt(1-vz^2)/vx = sqrt(vx^2)/vx = 1 for valleys and -1 for anti-valleys
+  // a=0
+    
+    
+  // We go on with "general case" where  vx=!0 and vy!=0
+    
+  // We start by computing b (if vx=!0 and vy!=0) : 
+  // f = vxb - vya
+  // f = vxb - vy (-bvy/vx)
+  // f = b(vx + vy^2/vx)
+  // f = b/vx(vx^2 + vy^2)
+  // b = fvx/(vx^2 + vy^2)
+  // b = fvx/(1 - vz^2)
+  // b = fvx/f^2
+  // b = vx/f
+    
+  // Then we compute a (if vx=!0 and vx!=0) : 
+  // a = -bvy/vx
+  // a = -vxvy/e/vx
+  // a = -vy/e
+    
+    
+  // Finally, we compute d and e which is the same for every case :
+  // d=-vzb
+  // e=vza
+    
+    
+  double vx=valleyDirVec.x();
+  double vy=valleyDirVec.y();
+  double vz=valleyDirVec.z();
+  double normval=sqrt(vx*vx+vy*vy+vz*vz); 
+  vx=vx/normval;
+  vy=vy/normval;
+  vz=vz/normval;
+  double a;
+  double b;
+  double c=0; // rotated y axis must stay in the X-Y plane
+  double d;
+  double e;
+  double f;
+  double antia; // find the corresponding anti-valley rotation matrix elements
```

Comment:
You should not need to do the anti-valley stuff in parallel.  You should be able to call this function with the anti-valley direction and let it do the same work.